url294e6c45f3_Ecdosis_NMergeNew_tgz-pJ8-TestJ8/src/edu/luc/nmerge/fastme/tree.java

```diff
@@ -0,0 +1,1616 @@
+		   		}
+		   		if (null != par)
+		   		{
+		   			if (e.tail.leftEdge == e)
+		   				updateSubTreeAverages(A, par, v, direction.LEFT);
+		   			else
+		   				updateSubTreeAverages(A, par, v, direction.RIGHT);
+		   			A[right.head.index][par.head.index]
+		   			     = A[par.head.index][right.head.index]
+		   			     = (right.bottomsize*A[right.head.index][par.head.index]
+		   			     + A[v.index][par.head.index]) / (right.bottomsize + 1);
+		   		}
+		   		break;
+		}
+	}
+	private void GMEsplitEdge( node v, edge e, double[][] A )
+	{
+	  String nodelabel;
+	  String edgelabel;
+	  edge newPendantEdge;
+	  edge newInternalEdge;
+	  node newNode;
+	    
+	  nodelabel = "I"+(this.size+1); 
+	  newNode = new node( nodelabel, null, this.size + 1 );  
+	  edgelabel = "E"+this.size; 
+	  newPendantEdge = new edge( edgelabel, newNode, v, 0.0 );   
+	  edgelabel = "E"+(this.size+1); 
+	  newInternalEdge = new edge( edgelabel, newNode, e.head, 0.0 );     
+	  //update the matrix of average distances
+	  //also updates the bottomsize, topsize fields
+	  
+	  GMEupdateAveragesMatrix(A,e,v,newNode);
+
+	  newNode.parentEdge = e;
+	  e.head.parentEdge = newInternalEdge;
+	  v.parentEdge = newPendantEdge;
+	  e.head = newNode;
+	  
+	  this.size = this.size + 2;
+
+	  if (e.tail.leftEdge == e) 
+	    {
+	      newNode.leftEdge = newInternalEdge;
+	      newNode.rightEdge = newPendantEdge;
+	    }
+	  else
+	    {
+	      newNode.leftEdge = newInternalEdge;
+	      newNode.rightEdge = newPendantEdge;
+	    }
+	  
+	  /*assign proper topsize, bottomsize values to the two new Edges*/
+	  
+	  newPendantEdge.bottomsize = 1; 
+	  newPendantEdge.topsize = e.bottomsize + e.topsize;
+	  
+	  newInternalEdge.bottomsize = e.bottomsize;
+	  newInternalEdge.topsize = e.topsize;  /*off by one, but we adjust
+						    that below*/
+	  
+	  /*and increment these fields for all other edges*/
+	  newInternalEdge.updateSizes(direction.UP);
+	  e.updateSizes(direction.DOWN);
+	}
+	void updatePair( double[][] A, edge nearEdge, edge farEdge, node v,
+			node root, double dcoeff, direction d )
+	{
+	  	edge sib;
+	  	//the various cases refer to where the new vertex has
+	    //been inserted, in relation to the edge nearEdge
+	  	switch( d ) 
+	    {
+	    	case UP: 
+	    		//this case is called when v has been inserted above 
+	    		//or skew to farEdge
+	    		//do recursive calls first!
+	    		if ( null != farEdge.head.leftEdge )
+	    			updatePair( A,nearEdge,farEdge.head.leftEdge,v,root,dcoeff, 
+	    			    direction.UP );
+	    		if ( null != farEdge.head.rightEdge )
+	    			updatePair(A,nearEdge,farEdge.head.rightEdge,v,root,
+	    			    dcoeff,direction.UP);
+	    		A[farEdge.head.index][nearEdge.head.index] =
+	    			A[nearEdge.head.index][farEdge.head.index]
+	    	        = A[farEdge.head.index][nearEdge.head.index]
+		            + dcoeff*A[farEdge.head.index][v.index]
+		            - dcoeff*A[farEdge.head.index][root.index];
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 536:540, 549:553.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.